Update upgrade guide for app mode (#728)
canonical/traefik-k8s-operator: docs/how-to/upgrade.md

```
juju refresh traefik-k8s
```

## Revision-specific upgrade notes

Some revisions require additional manual steps after upgrading. Check the section that applies to your current revision before refreshing.

(upgrade_rev281_to_rev308)=

> ### Upgrading from rev281–rev307 to rev308 or later

Reviewer: I think this should be upgrading from 280

Author: I mentioned below that eventhough we shipped app mode in rev 280, it's not until rev 308 that the unit-scoped secret issue was fixed, so I'd probably update this to say <308

Revision 308 fixes a bug where Traefik could request a new TLS certificate unnecessarily after a Juju leader change. If your deployment was running any revision between 281 and 307, follow these steps.

#### Background

> Revisions 281–307 stored the TLS private key in a **unit-scoped** Juju secret. Because each unit held its own private key, only the leader unit's key was used to generate the Certificate Signing Request (CSR). When the leader changed, the new leader had a different private key that did not match the existing certificate, causing an unnecessary certificate re-request.

Reviewer: Any revision< 281 had unit-scoped secret.

Author: I think it's <308 because eventhough we were already in APP mode the private key itself was still an unit-scoped secret because the library was outdated and had a bug. We bumped that lib in rev 308 iirc

> Revision 308 corrects this by storing the private key in a single **application-scoped** secret shared across all units. However, when upgrading from a previously affected revision, the old unit-scoped secrets are not automatically removed by Juju. Because the charm identifies secrets by label and `juju secret get` matches both unit-scoped and application-scoped secrets, the charm may find the stale unit-scoped secret and never create the correct application-scoped one.

Reviewer: Isn't revision 281 associated with https://github.com/canonical/traefik-k8s-operator/pull/620 already sharing a app scoped secret across all units?

Author: Yep, though a bug was in the lib that makes the charm still creates unit-scoped secrets. This was fixed in rev 308 iirc

> Revision 308 corrects this by storing the private key in a single **application-scoped** secret shared across all units. However, when upgrading from a previously affected revision, the old unit-scoped secrets are not automatically removed by Juju. Because the charm identifies secrets by label and `juju secret get` matches both unit-scoped and application-scoped secrets, the charm may find the stale unit-scoped secret and never create the correct application-scoped one.

Reviewer: > Because the charm identifies secrets by label and `juju secret get` matches both unit-scoped and application-scoped secrets

The TLS library looks for the Mode and correspondingly fetches either the unit or the app secret only, not both

Author: I think the mode only matters when creating the secret, fetching them is done by `self.charm.model.get_secret(label=...)` and that can potentially return either the app-scoped one or the unit-scoped one, if there's still the unit-scope one leftover from a previous revision

#### Steps

After running `juju refresh traefik-k8s --revision 308` (or later), check whether stale unit-scoped secrets remain:

```bash
juju secrets --format json | jq '.[] | select(.label | test("private-key"; "i"))'
```

> If the output shows secrets with unit owners (e.g. `traefik-k8s/0`, `traefik-k8s/1`), the stale secrets are present and must be cleaned up. Proceed with listing and deleting each unit-scoped private-key secret:

Reviewer 2: Shouldn't we recommend to backup the content somewhere first? I know they are not supposed to be used, but we never now.

Author: updated

```bash
juju secrets --format json \
  | jq -r '.[] | select(.label | test("private-key"; "i")) | select(.owner | test("/")) | .id' \
  | xargs -I{} juju remove-secret {}

> ```

Reviewer: We will also see the CSRs in the unit databag which must be cleaned up, we need to think about how to clean them.

```bash
jhack fire traefik-k8s/leader config-changed
```

#### Verification

After either option, confirm that the private-key secret is now application-scoped:

```bash
juju secrets --format json \
  | jq '.[] | select(.label | test("private-key"; "i")) | {id, label, owner}'
```

The `owner` field should show the application name (e.g. `traefik-k8s`) rather than a unit name. The charm will use this shared key for all CSRs going forward, and leader changes will no longer trigger unnecessary certificate renewals.

## Preserving TLS certificates after upgrade

By default, `juju refresh` to a revision that uses APP mode will cause Traefik to generate a new private key, which triggers a new Certificate Signing Request (CSR). If you are using `manual-tls-certificates` or `notary` as your TLS provider, this means you will need to sign and provide a new certificate.

To avoid this, you can restore the original private key immediately after the refresh, so that Traefik re-uses the same key and produces the same CSR that was already signed. This allows you to provide the original certificate without any re-signing.

### Steps

> **1. Before upgrading**, retrieve and save the current private key and certificate from any Traefik unit:

Reviewer 2: As people may run the documentation "after upgrading", maybe we should add a not to mention that the key/crt may just be available on one/some units?

Author: added!

> ```

Reviewer 2: ```suggestion
```bash
```

juju ssh --container traefik traefik-k8s/0 cat /opt/traefik/juju/<hostname>.key
juju ssh --container traefik traefik-k8s/0 cat /opt/traefik/juju/<hostname>.crt

> ```

Reviewer: What happens if the user has already upgraded and lost the old private key and cert? Its better to put this step in the very beginning of the document before asking them to upgrade.

Author: @copilot move this to the very first step before upgrade

> **2. Restore the private key on the leader unit.**

Reviewer 2: To be discussed if we turn this into an action.

On the `charm` container of the leader unit, create a temporary helper in `/var/lib/juju/agent/unit-<>/charm/src/charm.py` by appending the following method to the `TraefikIngressCharm` class. Replace the placeholder key with the one you saved in step 1:

```python
def _update_private_key(self) -> None:
    import textwrap
    private_key = textwrap.dedent("""-----BEGIN RSA PRIVATE KEY-----
<paste your private key here>
-----END RSA PRIVATE KEY-----""")
    secret = self.model.get_secret(label=self.certs._get_private_key_secret_label())
    secret.set_content({"private-key": str(private_key)})
```

You can then run this method on the leader unit using `jhack`:

```
jhack eval traefik-k8s/leader self._update_private_key()
```

**3. Fire a `config-changed` event** to trigger Traefik to re-generate the CSR with the restored key:

```
jhack fire traefik-k8s/leader config-changed
```

> After this, the original CSR will reappear in the relation databag. You can then provide the previously signed certificate without any re-signing:

Reviewer: As mentioned in the standup, how can we be sure if the same CSR is generated? If things like datetime factor into CSR creation then this will break.

Reviewer: In fact, looking at the certificate library, I see a fresh unique subject identifier is added by default every time CSR generation happens. I'm surprised this worked on sunbeam.

Author: I'll probably need to update the wording on this, it's not "the original CSR" but "a CSR created by the original private key"

Author: @copilot update "the original CSR" to "a CSR created by the original private key"

Reviewer: @Thanhphan1147  But a new CSR would require resigning correct?

Author: discussed internally, it won't require a resigning

Reviewer: it doesn't require resigning. As long as the private key and cert are consistent, manual tls charm can just feed the old certificate back to traefik with the new CSR.

```
juju refresh traefik-k8s
```

## Revision-specific upgrade notes

Some revisions require additional manual steps after upgrading. Check the section that applies to your current revision before refreshing.

(upgrade_to_rev308)=

> ### Upgrading to revision 308 or later

Reviewer 3: ```suggestion
### Upgrade to revision 308 or later
```

> Revision 308 officially switch the certificates management in Traefik from UNIT mode to APP mode. If your deployment was running any revision before 308, follow these steps.

Reviewer 3: ```suggestion
Revision 308 officially switches the certificates management in Traefik from UNIT mode to APP mode. If your deployment uses any revision before 308, follow these steps to preserve your TLS certificates.

This guide assumes that you have [`jhack`](https://snapcraft.io/jhack) installed.
```
Nits, I think we should be consistent in the tense that we use (in this case, I opted for present tense). I also think that we should be more explicit about what the steps will achieve.

Finally I added a note about `jhack`

> #### Important: Preserving TLS certificates

Reviewer 3: ```suggestion
#### Preserve TLS certificates
```
I'm not totally opposed to including `Important:` in the section header, but I do want to avoid using gerunds so that the header is more action-oriented.

If you think that we should include `Important` in the header, let's discuss more! I would love to understand why this section would benefit from including an `Important` flag.

> By default, `juju refresh` to a revision that uses APP mode will cause Traefik to generate a new private key, which triggers a new Certificate Signing Request (CSR). If you are using `manual-tls-certificates` or `notary` as your TLS provider, this means you will need to sign and provide a new certificate.

Reviewer 3: ```suggestion
By default, upgrading to a revision that uses APP mode will cause Traefik to generate a new private key, which triggers a new Certificate Signing Request (CSR). If you use `manual-tls-certificates` or `notary` as your TLS provider, this means you will need to sign and provide a new certificate.
```
nits

> To avoid this, you can restore the original private key immediately after the refresh, so that Traefik re-uses the same key and produces the same CSR that was already signed. This allows you to provide the original certificate without any re-signing.

Reviewer 3: ```suggestion
To avoid providing a new certificate, you can restore the original private key immediately after the refresh, so that Traefik re-uses the same key and produces the same CSR that was already signed. This allows you to provide the original certificate without any re-signing.
```
nit, to avoid a lone "this"

Before proceeding with the upgrade, save the existing private key and certificate **before** running `juju refresh`:

```bash
juju ssh --container traefik traefik-k8s/0 cat /opt/traefik/juju/<hostname>.key
juju ssh --container traefik traefik-k8s/0 cat /opt/traefik/juju/<hostname>.crt

> ```

Reviewer: instead of 0 i think we should specify the leader. The 0th unit might not be the right private key.

> Note: In some cases, a bug might occur during a node restart or a leader change that causes the leader unit to wipe out the certificates information in

> `/opt/traefik/juju`, you can look at other units to see if they still contain the original certificate and private key. In an HA deployment, you can also get them by looking at the `peer` relation data.

Reviewer 3: Let's use an admonition for this note (if you use GitHub's auto-incorporation feature, double check that the formatting is correct):
````suggestion
```{note}
In some cases, a bug might occur during a node restart or a leader change that causes the leader unit to wipe out the certificates information in
`/opt/traefik/juju`. Look at other units to see if they still contain the original certificate and private key. In an HA deployment, you can also get them by looking at the `peer` relation data.
```
````
I also added a nit to split up the longer sentence into two.

> `/opt/traefik/juju`, you can look at other units to see if they still contain the original certificate and private key. In an HA deployment, you can also get them by looking at the `peer` relation data.

Reviewer: how would they know which unit has the right private key? If they have 2 other units with different private key which one would they pick?

> #### Steps

Reviewer 3: ```suggestion
#### Update secrets
```
I prefer to avoid meta-documentation like "Steps" in favor of section headers that explicitly tell the user what actions they'll perform. Let me know if you think we should go with something other than `Update secrets`!

> After running `juju refresh traefik-k8s --revision 308` (or later), check whether stale unit-scoped secrets remain:

Reviewer: ```
swetha.swaminathan@canonical.com@zbook:~/Canonical/hrms-operator$(feat/frappe-hrms-charm)juju refresh traefik-k8s --revision 308
ERROR cannot upgrade from single base "ubuntu@20.04" charm to a charm supporting ["ubuntu@26.04"]. Use --force-series to override.
```
This fails and sunbeam is using amd64 so maybe we should lose the revision number

```bash
juju secrets --format json | jq '.[] | select(.label | test("private-key"; "i"))'
```

If the output shows secrets with unit owners (e.g. `traefik-k8s/0`, `traefik-k8s/1`), the stale secrets are present and must be cleaned up. Proceed with listing and deleting each unit-scoped private-key secret:

```bash
juju secrets --format json \
  | jq -r '.[] | select(.label | test("private-key"; "i")) | select(.owner | test("/")) | .id' \
  | xargs -I{} juju remove-secret {}
```

After deletion, trigger a reconciliation on the leader unit so the charm creates a new application-scoped secret:

```bash

> jhack fire traefik-k8s/leader config-changed

Reviewer 3: AFAIK `jhack` isn't shipped with Juju, so including this command means we're assuming that the user has already installed `jhack`. That's fine but we should explicitly state this in the guide. I added this to my suggestion for updating the introduction

> jhack fire traefik-k8s/leader config-changed

Reviewer: with my changes i expect the charm to go into ErrorState, if thats the case they can just run juju resolve and a new private key and CSR will be created.

```

> #### Verification

Reviewer 3: ```suggestion
#### Verify secrets
```

> After either option, confirm that the private-key secret is now application-scoped:

Reviewer 3: ```suggestion
Confirm that the private-key secret is now application-scoped:
```
nit for conciseness

```bash
juju secrets --format json \
  | jq '.[] | select(.label | test("private-key"; "i")) | {id, label, owner}'
```

The `owner` field should show the application name (e.g. `traefik-k8s`) rather than a unit name. The charm will use this shared key for all certificate signing requests going forward, and leader changes will no longer trigger unnecessary certificate renewals.

> ## Preserving TLS certificates after upgrade

Reviewer 3: ```suggestion
## Preserve TLS certificates after upgrade
```

> Assuming that you've done the necessary steps to backup the private key and certificate mentioned above, you should be able to now restore the private key on the leader unit and "re-provide" the original signed certificates.

Reviewer 3: ```suggestion
Assuming that you've backed up the private key and certificate, you should be able to now restore the private key on the leader unit and re-provide the original signed certificates.
```
nits

### Steps

> **1. Restore the private key on the leader unit.**

Reviewer 3: ```suggestion
**1. Restore the private key on the leader unit.**
```
or
```suggestion
### Restore the private key on the leader unit.
```

Either way, I would like to avoid meta-documentation with the `Steps` header. If you prefer the second suggestion (promoting the `Restore the private key...` text as a header), then please do the same promotion for the other steps.

```
juju refresh traefik-k8s
```

> ## Revision-specific upgrade notes

Reviewer 3: ```suggestion
## Revision-specific upgrades
```
nit

```bash
juju ssh --container traefik traefik-k8s/0 cat /opt/traefik/juju/<hostname>.key
juju ssh --container traefik traefik-k8s/0 cat /opt/traefik/juju/<hostname>.crt
```

```{note}
In some cases, a bug might occur during a node restart or a leader change that causes the leader unit to wipe out the certificates information in
`/opt/traefik/juju`. Look at other units to see if they still contain the original certificate and private key. In an HA deployment, you can also get them by looking at the `peer` relation data.
```

#### Update secrets

After running `juju refresh traefik-k8s --revision 308` (or later), check whether stale unit-scoped secrets remain:

```bash
juju secrets --format json | jq '.[] | select(.label | test("private-key"; "i"))'
```

If the output shows secrets with unit owners (e.g. `traefik-k8s/0`, `traefik-k8s/1`), the stale secrets are present and must be cleaned up. Proceed with listing and deleting each unit-scoped private-key secret:

```bash
juju secrets --format json \
  | jq -r '.[] | select(.label | test("private-key"; "i")) | select(.owner | test("/")) | .id' \

>   | xargs -I{} juju remove-secret {}

Reviewer: btw this doesn't work. ID is not a field in the json, but rather they key itself:
```
swetha.swaminathan@canonical.com@zbook:~$juju secrets --format json \
  | jq -r '.[] | select(.label | test("private-key"; "i")) | select(.owner | test("/")) | .id' \
  | xargs -I{} juju remove-secret {}
ERROR secret "null" not found
ERROR secret "null" not found
ERROR secret "null" not found
swetha.swaminathan@canonical.com@zbook:~$juju secrets --format json | jq
{
  "122sdmbq59kfqvn2b750": {
    "revision": 1,
    "owner": "traefik/2",
    "label": "afd8c2bccf834997afce12c2706d2ede-private-key-certificates",
    "created": "2026-07-06T07:47:28Z",
    "updated": "2026-07-06T07:47:28Z"
  },
  "2j6il33nfla5ppnjaic0": {
    "revision": 1,
    "owner": "traefik/0",
    "label": "afd8c2bccf834997afce12c2706d2ede-private-key-certificates",
    "created": "2026-07-06T07:47:23Z",
    "updated": "2026-07-06T07:47:23Z"
  },
  "9ele4veiueitmufq2t9g": {
    "revision": 1,
    "expires": "2027-05-30T19:47:59Z",
    "owner": "traefik/0",
    "label": "afd8c2bccf834997afce12c2706d2ede-certificate-eaf987838610f9ee6358fafe07da7f799ae48ee4653ca48cd268dab32338b2c6",
    "created": "2026-07-06T07:48:11Z",
    "updated": "2026-07-06T07:48:11Z"
  },
  "9if672umgk0qdcb4m7og": {
    "revision": 1,
    "owner": "traefik",
    "label": "tls-key",
    "created": "2026-07-06T07:48:11Z",
    "updated": "2026-07-06T07:48:11Z"
  },
  "f7uoilbpb55duvn4c8fg": {
    "revision": 1,
    "owner": "traefik/1",
    "label": "afd8c2bccf834997afce12c2706d2ede-private-key-certificates",
    "created": "2026-07-06T07:47:29Z",
    "updated": "2026-07-06T07:47:29Z"
  }
}
swetha.swaminathan@canonical.com@zbook:~$juju secrets --format json \
| jq -r '
    to_entries[]
    | select(.value.label | test("private-key"; "i"))
    | select(.value.owner | test("/"))
    | .key
'
122sdmbq59kfqvn2b750
2j6il33nfla5ppnjaic0
f7uoilbpb55duvn4c8fg
```

Author: @copilot update the command to use `.key` instead if `.id`
